refactor: Extract split filter provider interface to improve flexibility for user customization. (#46)
y-scope/presto: presto-docs/src/main/sphinx/connector/clp.rst

> #############

Reviewer: To support a deeper nesting level, I refactored all the headings to follow the heading level syntax described [here](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections).

I also removed extra line breaks before headers.

                                     database name is not specified in the URL.
``clp.metadata-db-password``         The password for the metadata database user. This option is required if
                                     ``clp.metadata-provider-type`` is set to ``mysql``.
``clp.metadata-table-prefix``        A string prefix prepended to all metadata table names when querying the
                                     database. Useful for namespacing or avoiding collisions. This option is
                                     required if ``clp.metadata-provider-type`` is set to ``mysql``.
``clp.metadata-expire-interval``     Defines how long, in seconds, metadata entries remain valid before they  600
                                     need to be refreshed.
``clp.metadata-refresh-interval``    Specifies how frequently metadata is refreshed from the source, in       60
                                     seconds. Set this to a lower value for frequently changing datasets or
                                     to a higher value to reduce load.
``clp.split-filter-config``          The absolute path to an optional split filter config file. See the

>                                      :ref:`Split Filter Config File<split-filter-config-file>` section
>                                      for details.

Reviewer: ```suggestion
                                     :ref:`Split Filter Config File<split-filter-config-file>` section for
                                     details.
```

Filter Configs
^^^^^^^^^^^^^^

Each `filter config` indicates how a *data column*---a column in the Presto table, will be used as a filter to query in
the metadata database. The basic structure of a filter contains the following three fields:

- ``columnName``: indicates the data column's name.

- ``metadataDatabaseSpecific`` *(optional)*: stores fields specific to some metadata database that can be used for metadata filtering.
 An example for MySQL-specific fields can be seen
  :ref:`here<example-of-metadata-filters-for-mysql-metadata-database>`. This provides the ability to generate

> metadata filter queries for different types of metadata databases by modifying the pushed-down expression for metadata filtering according to the contents of this field.

Author: partially applied

- ``required`` *(optional, defaults to false)*: indicates whether the filter **must** be present in the pushed-down
  expression for metadata filtering. If a required filter is missing or cannot be pushed down, the query will be
  rejected.

For example, an integer data column (e.g., ``timestamp``), may be remapped to a pair of metadata columns that represent
the range of possible values (e.g., ``begin_timestamp`` and ``end_timestamp``) of the data column within a split.

.. _example-of-metadata-filters-for-mysql-metadata-database:

Example of Metadata Filters for MySQL Metadata Database

> ^^^^^^^

Author: applied

> The field ``metadataDatabaseSpecific`` has the following properties:

Author: applied

Filter Configs
^^^^^^^^^^^^^^

Each `filter config` indicates how a *data column*---a column in the Presto table, will be used as a filter to query in
the metadata database. The basic structure of a filter contains the following three fields:

- ``columnName``: indicates the data column's name.

- ``metadataDatabaseSpecific`` *(optional)*: stores the metadata-database-specific fields which can be used for metadata
  filtering. We will provide an example of the MySQL-specific
  :ref:`here<example-of-metadata-filters-for-mysql-metadata-database>`. This field provides the ability to generate the

>   metadata filter query for different types of metadata database from the pushed-down expression for metadata filtering.

Reviewer 2: ```suggestion
- ``metadataDatabaseSpecific`` *(optional)*: stores fields specific to some metadata database that can be used for metadata filtering.
 An example for MySQL-specific fields can be seen
  :ref:`here<example-of-metadata-filters-for-mysql-metadata-database>`. This provides the ability to generate
metadata filter queries for different types of metadata databases by modifying the pushed-down expression for metadata filtering according to the contents of this field.
```

Edited a bit for clarity based on my understanding of how this works. Feel free to change it up more though if I got something wrong or if there is a clearer way to write this.

> - ``required`` *(optional, defaults to false)*: indicates whether the filter **must** be present in the pushed-down

Reviewer 2: Just curious, when would we require a filter? Is it just for situations where it is so important for performance that we're better of erroring out if we can not push down a filter to the metadata database?

Author: yea it is mostly for performance concern. The filters are only used when querying the metadata database to retrieve splits to process. So if there are no filters, it will literally process all splits, which can take very long time if the data volumn is big.

  expression for metadata filtering. If a required filter is missing or cannot be pushed down, the query will be
  rejected.

For example, an integer data column (e.g., ``timestamp``), may be remapped to a pair of metadata columns that represent
the range of possible values (e.g., ``begin_timestamp`` and ``end_timestamp``) of the data column within a split.

.. _example-of-metadata-filters-for-mysql-metadata-database:

Example of Metadata Filters for MySQL Metadata Database
^^^^^^^

> For the filed ``metadataDatabaseSpecific``, it has the following field:

Reviewer 2: ```suggestion
The field ``metadataDatabaseSpecific`` has the following properties:
```

.. _filter-configs:

Filter Configs
==============

Each filter config indicates how a *data column*---i.e., a column in the Presto table---should be mapped to one or
more *metadata columns*---i.e., columns in CLP's metadata database.

For example, an integer data column (e.g., ``timestamp``), may be remapped to a pair of metadata columns that represent
the range of possible values (e.g., ``begin_timestamp`` and ``end_timestamp``) of the data column within a split.

Each filter config has the following options:

- ``columnName``: The data column's name.

> - ``customOptions`` *(optional)*: Custom config options for a split filter provider. Options for the default split
>   filter provider (``ClpMySqlSplitFilterProvider``) are :ref:`below<clp-mysql-split-filter-provider-config>`.

Reviewer: ```suggestion
- ``customOptions`` *(optional)*: Custom options for a split filter provider. Options for the default split filter
  provider (``ClpMySqlSplitFilterProvider``) are :ref:`below<clp-mysql-split-filter-provider-config>`.
```

> - ``required`` *(optional, defaults to false)*: Whether the filter **must** be present in the generated split filter
>   query. If a required filter is missing or cannot be added to the split filter query, the original query will be
>   rejected.

Reviewer: ```suggestion
- ``required`` *(optional, defaults to false)*: Whether the filter **must** be present in the generated metadata query.
  If a required filter is missing or cannot be added to the metadata query, the original query will be rejected.
```

.. _clp-mysql-split-filter-provider-config:

ClpMySqlSplitFilterProvider-Specific Filter Config
-----------------------------------------------

> For the ``ClpMySqlSplitFilterProvider``, the ``customOptions`` option of the filter config has the following
> sub-options:

Reviewer: ```suggestion
For ``ClpMySqlSplitFilterProvider``, the ``customOptions`` option of the filter config has the following sub-options:
```

  - ``lowerBound``: The metadata column that represents the lower bound of values in a split for the data column.
  - ``upperBound``: The metadata column that represents the upper bound of values in a split for the data column.

Filter Config Example
---------------------

> The code block shows an example split filter config file:

Reviewer: ```suggestion
The code block shows an example filter config file:
```

    ``end_timestamp``, and is required to exist in every query.
  - The column ``file_name`` is used as-is without remapping.

> If you prefer to use a different format for the ``customOptions``, you can provide your own implementations of the
> ``ClpSplitFilterProvider`` interface, and configure the connector accordingly.

Reviewer: ```suggestion
If you prefer to use a different format for ``customOptions``, you can provide your own implementation of the
``ClpSplitFilterProvider`` interface, and configure the connector accordingly.
```
